flow/aggregate.py

```diff
@@ -0,0 +1,183 @@
+        # copied from: https://docs.python.org/3/library/itertools.html#itertools.zip_longest
+        try:
+            num = int(num)
+            if num < 0:
+                raise ValueError("The num parameter should be greater than 0")
+        except Exception:
+            raise TypeError("The num parameter should be an integer")
+
+        def aggregator(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        return cls(aggregator, sort, reverse, select)
+
+    @classmethod
+    def groupby(cls, key, default=None, sort=None, reverse=False, select=None):
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+
+        elif isinstance(key, Iterable):
+            keys = list(key)
+
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                def keyfunction(job):
+                    return [job.sp.get(key, default) for key in keys]
+
+        elif callable(key):
+            keyfunction = key
+
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            "or a callable, got {}".format(type(key)))
+
+        def aggregator(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator, sort, reverse, select)
+
+    def __call__(self, obj=None):
+        if callable(obj):
+            setattr(obj, '_flow_aggregate', self)
+            return obj
+        elif isinstance(obj, (list, Project, JobsCursor)):
+            aggregated_jobs = deepcopy(list(obj))
+            if self._select is not None:
+                aggregated_jobs = list(self._select(aggregated_jobs))
+            if self._sort is not None:
+                aggregated_jobs = list(self._sort(aggregated_jobs))
+            aggregated_jobs = self._aggregator([job for job in aggregated_jobs])
+            aggregated_jobs = self._create_nested_aggregate_list(aggregated_jobs)
+            if not len(aggregated_jobs):
+                return []
+            for i, job in enumerate(aggregated_jobs[-1]):
+                if job is None:
+                    del aggregated_jobs[-1][i:]
+                    break
+            return aggregated_jobs
+        elif isinstance(obj, signac.contrib.job.Job):
+            # For backwards compatibility
+            return [obj]
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable '
+                            'or an Iterable of Jobs, got {}.'.format(type(obj)))
+
+    def _create_nested_aggregate_list(self, aggregated_jobs):
```

Comment:
This function is complicated quite a bit by allowing for iterators that return single jobs. I think we should not allow that behavior.